dftimewolf/lib/collectors/grr_hosts.py

```diff
@@ -980,64 +982,87 @@ def _DownloadResults(self,
+          data_frame=data_frame,
+          flow_identifier=flow_identifier,
+          client_identifier=client_identifier)
+
+      self.state.StoreContainer(dataframe_container)
+
   def Process(self, container: containers.Host) -> None:
     """Collect osquery results from a host with GRR.
 
     Raises:
       DFTimewolfError: if no artifacts specified nor resolved by platform.
     """
-    for client in self._FindClients([container.hostname]):
-      osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
-
-      for osquery_container in osquery_containers:
-        hunt_args = osquery_flows.OsqueryFlowArgs(
-            query=osquery_container.query,
-            timeout_millis=self.timeout_millis,
-            ignore_stderr_errors=self.ignore_stderr_errors)
-
-        try:
-          flow_id = self._LaunchFlow(client, 'OsqueryFlow', hunt_args)
-          self._AwaitFlow(client, flow_id)
-        except DFTimewolfError as error:
-          self.ModuleError(
-              f'Error raised while launching/awaiting flow: {error.message}')
-          continue
+    client = self._GetClientBySelector(container.hostname)
 
-        name = osquery_container.name
-        description = osquery_container.description
-        query = osquery_container.query
-        hostname = container.hostname
-        flow_identifier = flow_id
-        client_identifier = client.client_id
-
-        results = self._DownloadResults(client, flow_id)
-        if not results:
-          results_container = containers.OsqueryResult(
-              name=name,
-              description=description,
-              query=query,
-              hostname=hostname,
-              data_frame=pd.DataFrame(),
-              flow_identifier=flow_identifier,
-              client_identifier=client_identifier)
-          self.state.StoreContainer(results_container)
-          continue
+    osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
 
-        for data_frame in results:
-          self.logger.info(
-              f'{str(flow_id)} ({container.hostname}): {len(data_frame)} rows '
-              'collected')
-
-          dataframe_container = containers.OsqueryResult(
-              name=name,
-              description=description,
-              query=query,
-              hostname=hostname,
-              data_frame=data_frame,
-              flow_identifier=flow_identifier,
-              client_identifier=client_identifier)
+    host_osquery_futures = []
+    with ThreadPoolExecutor(self.GetQueryThreadPoolSize()) as executor:
+      for osquery_container in osquery_containers:
+        host_osquery_future = executor.submit(
+          self._ProcessQuery, container.hostname, client, osquery_container)
+        host_osquery_futures.append(host_osquery_future)
 
-          self.state.StoreContainer(dataframe_container)
+    for host_osquery_future in host_osquery_futures:
+      if host_osquery_future.exception():
```

Comment:
How will this behave with `self.ModuleError`? Maybe this was your idea from the start, to use DFTimewolfExceptions in `_ProcessQuery` so that they would get caught here?

Reply:
I would also add a check that if ALL threads errored out, then we do a critical `ModuleError` as there's probably no point in continuing the recipe.

Reply:
> How will this behave with self.ModuleError? Maybe this was your idea from the start, to use DFTimewolfExceptions in _ProcessQuery so that they would get caught here?

This should catch any "critical" ModuleErrors as well as any other unhandled exceptions raised in the worker.   I don't think it was my original intention to do that here - it was more to handle any unknown thread related issues that might arise during execution.

> I would also add a check that if ALL threads errored out, then we do a critical ModuleError as there's probably no point in continuing the recipe.

I'm assuming you mean all threads for a single host in which case I think it would be OK to continue the other threads processing other hosts?   Main scenario that comes to mind when all threads would fail would be if there are timeout issues due to the host being unresponsive/offline.

Reply:
Yeah that's a fair point. Thanks for clarifying.